Proofs/Experiments/Max/MaxTandem.lean

```diff
@@ -0,0 +1,899 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Shilpi Goel
+
+Experimental: Use the Correctness module to prove that this program computes the
+following via a naive loop that iterates over `x0`:
+`x0 := x0 + x1`
+-/
+import Arm
+import Tactics.CSE
+import Tactics.Sym
+import Tactics.StepThms
+import Correctness.ArmSpec
+import Proofs.Experiments.Max.MaxProgram
+import Arm.Insts.Common
+import Arm.Memory.SeparateAutomation
+
+namespace MaxTandem
+open Max
+
+
+section PC
+
+/--# We define scoped notation for our cutpoint PCs to use in pattern matching. -/
+
+scoped notation "entry_start" => 0x894#64
+scoped notation "then_start" => 0x8b0#64
+scoped notation "else_start" => 0x8bc#64
+scoped notation "merge_end" => 0x8cc#64
+
+
+namespace ArmStateNotation
+
+/-! We build a notation for `read_mem_bytes $n $base $s` as `$s[$base, $n]` -/
+@[inherit_doc read_mem_bytes]
+syntax:max term noWs "[" withoutPosition(term)  ","  withoutPosition(term) noWs "]" : term
+macro_rules | `($s[$base,$n]) => `(read_mem_bytes $n $base $s)
+end ArmStateNotation
+
+open ArmStateNotation
+
+def pcs : List (BitVec 64) := [
+  entry_start,
+  then_start,
+  else_start,
+  merge_end
+  ]
+
+end PC
+
+def cut (s : ArmState) : Bool := decide (read_pc s ∈ pcs)
+
+section Invariants
+
+/-- Precondition for the correctness of the Add program. -/
+def pre (s : ArmState) : Prop :=
+  read_pc s = entry_start ∧
+  s.program = program ∧
+  read_err s = StateError.None ∧
+  -- (FIXME) We don't really need the stack pointer to be aligned, but the
+  -- `sym_n` tactic expects this. Can we make this optional?
+  CheckSPAlignment s ∧
+  s.sp ≥ 32 ∧
+  mem_legal' s.sp 80 -- TODO: find the correct smallest bound we need here.
+
+/-- Specification function. -/
+def spec (x0 x1 : BitVec 64) : BitVec 64 :=
+  if BitVec.slt (x0.zeroExtend 32)  (x1.zeroExtend 32)
+  then (x1.zeroExtend 32).zeroExtend 64
+  else (x0.zeroExtend 32).zeroExtend 64
+
+def post (s0 sf : ArmState) : Prop :=
+  sf.x0 = BitVec.zeroExtend 64 (BitVec.zeroExtend 32 (spec s0.x0 s0.x1)) ∧
+  read_pc sf = 0x8cc#64 ∧
+  read_err sf = StateError.None ∧
+  sf.program = program ∧
+  -- (FIXME) We don't really need the stack pointer to be aligned, but the
+  -- `sym_n` tactic expects this. Can we make this optional?
+  CheckSPAlignment sf
+
+def entry_start_inv (s0 si : ArmState) : Prop := si = s0
+
+/-- Invariant right at the conditional branch -/
+def entry_end_inv (s0 si : ArmState): Prop :=
+  read_err si = .None ∧
+  si.program = program ∧
+  -- (FIXME) We don't really need the stack pointer to be aligned, but the
+  -- `sym_n` tactic expects this. Can we make this optional?
+  CheckSPAlignment si ∧
+  si[s0.sp - 32#64 + 12#64, 4] = s0.x0.zeroExtend 32 ∧
+  si[s0.sp - 32#64 + 8#64, 4] = s0.x1.zeroExtend 32 ∧
```

Comment:
There is a function called `read_gpr` that takes a `width` argument. That function will do `zeroExtend` to the `width`. I recommend using it to hide the `zeroExtend`s so that it is easier to read.

Reply:
@pennyannn on the contrary, I prefer this, because it makes it clear that we're reading from `x0` /`x1`. I'm happy to have an alternative that name that hides the zero extend, but I'd like to keep the `s0.x0` / `s0.x1` for the massive boost in readability :)

Reply:
Could we invent new notations like `s0.w0` (Arm notation) which will stand for `s0.x1.zeroExtend 32`?

Reply:
At first I find the `zeroExtend`s a bit distracting when reading the proof, and I knew we have an abstraction `read_gpr` for hiding it. However, as I read more into the bulk of correctness proof, I found that you have lots of reference to `zeroExtend` in later proofs as well. So now I'm not sure if you needed to expand and expose `zeroExtend` anyways. I will let you decide if more abstraction could be done then...

Reply:
If we are building notations, then I'd prefer defining `s0.w0` and the like to stay true to Arm.